src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseMessage.cs

```diff
@@ -319,46 +332,75 @@ private static byte[] DecodeSignature(CborReader reader)
+            return signatures;
+        }
 
-            for (int i = 0; i < signaturesLength; i++)
+        private static CoseSignature DecodeCoseSignature(ReadOnlyMemory<byte> coseSignature)
+        {
+            var reader = new CborReader(coseSignature);
+            int? length = reader.ReadStartArray();
+
+            if (length.HasValue ? length != CoseMultiSignMessage.CoseSignatureArrayLength :
+                HasIndefiniteLengthArrayIncorrectLength(coseSignature, CoseMultiSignMessage.CoseSignatureArrayLength))
             {
-                int? length = reader.ReadStartArray();
+                throw new CryptographicException(SR.Format(SR.DecodeErrorWhileDecoding, SR.DecodeCoseSignatureMustBeArrayOfThree));
+            }
 
-                if (length != CoseMultiSignMessage.CoseSignatureArrayLength)
-                {
-                    throw new CryptographicException(SR.Format(SR.DecodeErrorWhileDecoding, SR.DecodeCoseSignatureMustBeArrayOfThree));
-                }
+            var protectedHeaders = new CoseHeaderMap();
+            DecodeProtectedBucket(reader, protectedHeaders, out byte[] signProtected);
 
-                var protectedHeaders = new CoseHeaderMap();
-                DecodeProtectedBucket(reader, protectedHeaders, out byte[] signProtected);
+            var unprotectedHeaders = new CoseHeaderMap();
+            DecodeUnprotectedBucket(reader, unprotectedHeaders);
 
-                var unprotectedHeaders = new CoseHeaderMap();
-                DecodeUnprotectedBucket(reader, unprotectedHeaders);
+            if (ContainDuplicateLabels(protectedHeaders, unprotectedHeaders))
+            {
+                throw new CryptographicException(SR.Sign1SignHeaderDuplicateLabels);
+            }
 
-                if (ContainDuplicateLabels(protectedHeaders, unprotectedHeaders))
-                {
-                    throw new CryptographicException(SR.Sign1SignHeaderDuplicateLabels);
-                }
+            byte[] signatureBytes = DecodeSignature(reader);
+            reader.ReadEndArray();
 
-                byte[] signatureBytes = DecodeSignature(reader);
+            return new CoseSignature(protectedHeaders, unprotectedHeaders, signProtected, signatureBytes);
+        }
 
-                signatures.Add(new CoseSignature(protectedHeaders, unprotectedHeaders, bodyProtected, signProtected, signatureBytes));
+        private static bool HasIndefiniteLengthArrayIncorrectLength(ReadOnlyMemory<byte> encodedArray, int expectedLength)
+        {
+            var reader = new CborReader(encodedArray);
+            reader.ReadStartArray();
+            int count = 0;
 
-                reader.ReadEndArray();
+            while (reader.PeekState() != CborReaderState.EndArray)
+            {
+                reader.SkipValue();
+                count++;
+
+                if (count > expectedLength)
+                    return true;
```

Comment:
The style rules say

> A single line statement block can go without braces but the block must be properly indented on its own line and _must not be nested in other statement blocks that use braces_

(emphasis mine)

So this should be

```suggestion
                if (count > expectedLength)
                {
                    return true;
                }
```